transport/grpc/dialer.go

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package grpc
+
+import (
+	"go.uber.org/yarpc/api/peer"
+)
+
+// NewDialer creates a transport that is decorated to retain peers with
+// additional gRPC dial options.
+func (t *Transport) NewDialer(options ...DialOption) *Dialer {
+	return &Dialer{trans: t, options: newDialOptions(options)}
+}
+
+// Dialer is a decorator for a gRPC transport that threads dial options for
+// every retained peer.
+type Dialer struct {
+	trans   *Transport
+	options *dialOptions
+}
+
+var _ peer.Transport = (*Dialer)(nil)
+
+// RetainPeer retains the identified peer with dial options.
+func (d *Dialer) RetainPeer(id peer.Identifier, ps peer.Subscriber) (peer.Peer, error) {
+	return d.trans.retainPeer(id, d.options, ps)
+}
+
+// ReleasePeer releases a TLS peer.
```

Comment:
This comment is not parallel with `RetainPeer`.
What ensures that only "TLS peer"s will be released?

Reply:
These methods are implemented on the Dialer peer.Transport which has the TLS config as a dialOption and is used to create the peerChooser which gets attached to the outbound. This struct also paves the way for adding support for additional grpcOptions in the future.

Reply:
I also found this comment a little confusing (after @jkanywhere pointed it out) because nothing else in this file associates a Dialer specifically with TLS.  Can a Dialer be used for connections that are non-TLS?

Reply:
The dialer can be used for any set of dial options, including no extra dial options. At the moment, we only support a credentials dial option. @ztrain we should generalize this comment.

Reply:
Excellent. Thanks @bradbl for translating my confusion :)

Reply:
I will add more docs to dialer.go, including clarifying this comment.